java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java

```diff
@@ -0,0 +1,1585 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.vector;
+
+import static org.apache.arrow.memory.util.LargeMemoryUtil.capAtMaxInt;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import org.apache.arrow.memory.ArrowBuf;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.memory.OutOfMemoryException;
+import org.apache.arrow.memory.util.ArrowBufPointer;
+import org.apache.arrow.memory.util.ByteFunctionHelpers;
+import org.apache.arrow.memory.util.CommonUtil;
+import org.apache.arrow.memory.util.hash.ArrowBufHasher;
+import org.apache.arrow.util.Preconditions;
+import org.apache.arrow.vector.compare.VectorVisitor;
+import org.apache.arrow.vector.ipc.message.ArrowFieldNode;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.apache.arrow.vector.util.CallBack;
+import org.apache.arrow.vector.util.OversizedAllocationException;
+import org.apache.arrow.vector.util.TransferPair;
+
+/**
+ * BaseVariableWidthViewVector is a base class providing functionality for strings/bytes types in view format.
+ *
+ */
+public abstract class BaseVariableWidthViewVector extends AbstractVariableWidthVector {
+  private static final int DEFAULT_RECORD_BYTE_COUNT = 16;
+  private static final int INITIAL_BYTE_COUNT = INITIAL_VALUE_ALLOCATION * DEFAULT_RECORD_BYTE_COUNT;
+  private static final int MAX_BUFFER_SIZE = (int) Math.min(MAX_ALLOCATION_SIZE, Integer.MAX_VALUE);
+  private int lastValueCapacity;
+  private long lastValueAllocationSizeInBytes;
+
+  public static final int OFFSET_WIDTH = 4; /* 4 byte unsigned int to track offsets */
+  /* protected members */
+  protected static final int INLINE_SIZE = 12; /* 12 byte unsigned int to track inline views*/
+  protected static final int VIEW_BUFFER_SIZE = 16; /* 16 byte default size for each view*/
+  protected static final int LENGTH_WIDTH = 4; /* the first 4 bytes of view are allocated for length*/
+  protected static final int PREFIX_WIDTH = 4; /* the second 4 bytes of view are allocated for prefix width*/
+  protected static final int BUF_INDEX_WIDTH = 4; /* third 4 bytes of view are allocated for buffer index*/
+  protected static final byte[] emptyByteArray = new byte[]{};
+  protected ArrowBuf validityBuffer;
+  protected ArrowBuf valueBuffer;
```

Comment:
We have more than one values buffer though?

Reply:
I think I should rename this. But before that let me explain the way I implemented the views. Ignoring the incorrect offset, the rest is as follows. 

The valueBuffer is basically the views.
It contains short string inline and long strings with the length, prefix, buf index and offset. The list of ArrowBuf a.k.a dataBuffers are the buffers which keeps the full strings for long strings. This buffer has a default size. Once it's filled to a certain limit where next long string cannot be stored a new ArrowBuf is added to this list. So the first element of the dataBuffer would be index 0 and beyond that 1, and then 2 after adding the third buffer. 

Is this incorrect? Is there a better alternative?

Reply:
As long as it's clearly documented.

Reply:
But as noted the spec calls it the "views" buffer.

Reply:
Of course my mistake for not properly naming and commenting. Is the way I used follow the spec? Sorry about the confusion.

Reply:
Renamed.